[Button][base] Drop `component` prop

mnajdova opened the pull request:
This PR removes the `component` prop from the `ButtonUnstyled` component. The motivation for this is that there is a duplicated API for setting the root slot (through the `component` prop and through the `slots.root` prop). The difference between these two APIs is that the `component` prop makes the component polymorphic, which means that you can send the root component's props on the `ButtonUnstyled` component directly, without getting typescript issues. However, this has shown to have a big performance implications when doing the type checks (see some [existing issues](https://github.com/mui/material-ui/issues?q=is%3Aissue+label%3Atypescript+label%3Aperformance+) for the Material UI's components). One of https://github.com/mui/material-ui/issues/36679

# BREAKING CHANGE

It will be covered by codemod, see https://github.com/mui/material-ui/pull/36831

For Base UI components,  the `component` prop value should be moved to `slots.root`: 

```diff
 <ButtonUnstyled
-  component="span"
+  slots={{ root: "span" }}
 />
```

If using TypeScript, you should add the custom component type as a generic on the `ButtonUnstyled` component.

```diff
-<ButtonUnstyled
+<ButtonUnstyled<typeof CustomComponent>
   slots={{ root: CustomComponent }}
   customProp="foo"
 />
```

michaldudak commented on packages/mui-base/src/ButtonUnstyled/ButtonUnstyled.types.ts:
```diff
@@ -43,11 +43,7 @@ export interface ButtonUnstyledSlots {
-> = OverrideProps<ButtonUnstyledTypeMap<{}, D>, D> & {
-  component?: D;
-};
+export type ButtonUnstyledProps = DefaultComponentProps<ButtonUnstyledTypeMap>;
```
I don't think we need `TypeMap` interfaces anymore. It could be simpler to use just the props types directly.

michaldudak replied:
Could we just explicitly define `ButtonUnstyledProps` as `ButtonUnstyledOwnProps & React.ButtonHTMLAttributes<HTMLButtonElement>`? This way readers see at first glance what are the props and we don't need to use the expensive `React.ComponentPropsWithRef` that's in `DefaultComponentProps`

michaldudak replied:
Also it makes me wonder if accepting all the native button props is correct in this case. Since the component is not polymorphic, supplying an `a` to `slots.root` while still using some button-specific attributes on the parent could lead to errors. Shouldn't we accept only the common HTML attributes?

mnajdova replied:
> Shouldn't we accept only the common HTML attributes?

I think this is too strict. If we go this way, we would need to then add and handle all of the props from the `ButtonHTMLAttributes` interface inside the component:

```ts
autoFocus?: boolean | undefined;
disabled?: boolean | undefined;
form?: string | undefined;
formAction?: string | undefined;
formEncType?: string | undefined;
formMethod?: string | undefined;
formNoValidate?: boolean | undefined;
formTarget?: string | undefined;
name?: string | undefined;
type?: 'submit' | 'reset' | 'button' | undefined;
value?: string | ReadonlyArray<string> | number | undefined;
```

I would go with your first proposal: `ButtonUnstyledOwnProps & React.ButtonHTMLAttributes<HTMLButtonElement>`, and if your second concern becomes a problem (for e.g. we see issues being created about it), we can add runtime warning for using these props - maybe after the component mounts we can check the host element and warn if it is not a `button` and some of the button specific attributes is used.

michaldudak replied:
In this case it could be worth creating a LinkButton component and removing the logic to render an `<a>` when `href` is present. I think the Button may be the only component used as frequently with the default component as with the `a` tag.

The props could also be typed as `(ButtonUnstyledWithoutHrefProps &  React.ButtonHTMLAttributes<HTMLButtonElement>) | (ButtonUnstyledWithHrefProps & React.AnchorHTMLAttributes<HTMLAnchorElement>)` - this way the typings will be correct for the most popular cases.

michaldudak replied:
Let's not make this discussion a blocker for other components, though. I think this issue is present only in a ButtonUnstyled (and perhaps in InputUnstyled to some extent). All the other components should work just fine with the default component props.

mnajdova replied:
This was basically solved with the generic component, I will create an issue for adding a demo for the ButtonUnstyled component used as a Link.

mnajdova commented on test/utils/describeConformanceUnstyled.tsx:
```diff
@@ -165,6 +165,10 @@ function testSlotsProp(element: React.ReactElement, getOptions: () => UnstyledCo
   });
 
   it('uses the component provided in the `component` prop when both `component` and `slots.root` are provided', () => {
+    if (skip && skip.indexOf('componentProp') >= 0) {
```
This test should be run only if we are not skipping the `component` prop test.

mnajdova commented on packages/mui-base/src/utils/OverridableComponent.ts:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @template TypeMap The interface the defines the props and a default root element of the component.
+ */
+export type OverridableComponent<TypeMap extends OverridableTypeMap> = {
```
```suggestion
export type PolymorphicComponent<TypeMap extends OverridableTypeMap> = {
```

@michaldudak what do you think about changing the name, so that there is no confusion with the two other implementations of it, especially as it is not identical?

michaldudak replied:
I don't really have a preference here. `PolymorphicComponent` does sound good, though - it communicates more clearly the intent.

mnajdova replied:
Done!

mnajdova commented on packages/mui-base/src/ButtonUnstyled/ButtonUnstyled.tsx:
```diff
@@ -143,34 +131,10 @@ ButtonUnstyled.propTypes /* remove-proptypes */ = {
-  /**
-   * @ignore
-   */
-  onBlur: PropTypes.func,
```
@michaldudak I think we should go the other way around and explicitly define these props in the ButtonUnstyledProps interface, as if a custom component is used, that does not have these props in the props interface, there will be a typescript error, which would be a bug in my opinion.

michaldudak replied:
I'm not sure if I understand you correctly. Why would there be a TS error? We don't actually use these props anywhere. Also in ButtonUnstyled.spec.tsx we have a case with `CustomComponent` that don't have these defined and it works fine.

mnajdova replied:
We don't have with custom component. For e.g. this is failing:

```tsx
<ButtonUnstyled<typeof CustomComponent>
  slots={{ root: CustomComponent }}
  stringProp="test"
  numberProp={0}
  onClick={() => {}}
/>
```

michaldudak replied:
OK, I get it now. It's an interesting problem. The `ButtonUnstyled` doesn't handle (or require) the click (or several other) events, so strictly speaking, it's unnecessary. It's necessary in the `slots.root`, though, but there is no way of enforcing it.
In other words, even if we had `onClick` in `ButtonUnstyledProps`, it would be possible to provide `CustomComponent` that ignores its `onClick` prop and effectively create a broken button.

mnajdova replied:
Here is a failing test: https://app.circleci.com/pipelines/github/mui/material-ui/95773/workflows/1ec2c63d-7103-4340-b894-909bf873b7a0/jobs/509209 

I think we should update the types to add these event handlers.

michaldudak replied:
A couple of thoughts I have about this:

1. How do we decide which events to define explicitly?
2. If they are defined explicitly, we will lose polymorphism in their parameters. We'll have to type them as either `MouseEventHandler<Element>` or `MouseEventHandler<HTMLButtonElement>`. In either case, another test will fail (the last one from the .spec suite).
3. If we use this pattern, we should also update all the other components

mnajdova replied:
I see your point. I've added ts-expect-error directive on the new test, so that in the future, if we decide to handle this, we have some record that it was an intended decision. We can discuss it separately tough in the future.

michaldudak commented on packages/mui-base/src/ButtonUnstyled/ButtonUnstyled.tsx:
I wonder if the ButtonUnstyled definition shouldn't be `function ButtonUnstyled(props: ButtonUnstyledProps, forwardedRef: React.ForwardedRef<any>)` (without the generics). This way, we get strongly typed props inside the component definition. Nothing would change with the exported definition as we cast it to PloymorphicComponent anyway.

mnajdova replied:
I like the explicitness that we have now. We can open a separate issue for this and see what other things I guess. It's anyway internal change

mnajdova commented on packages/mui-base/src/ButtonUnstyled/ButtonUnstyled.spec.tsx:
```diff
@@ -29,24 +29,28 @@ const polymorphicComponentTest = () => {
         onClick={(e: React.MouseEvent<HTMLButtonElement>) => e.currentTarget.checkValidity()}
       />
 
       <ButtonUnstyled<'div'>
```
```suggestion
      <ButtonUnstyled<'div'>
         slotProps={{ root: 'div' }}
```

samuelsycamore commented:
@mnajdova I renamed the `## Slot props` section to `## Custom structure` and updated the text. I'm not married to the header title, so I'm open to suggestions. (This would need to be implemented across the rest of the Base docs as well, some of which also discuss `slotProps` in this section. But it feels confusing to talk about "the slot prop" and "the slotProps prop" in a section called "Slot props." 😅)

I also updated the `Overriding component structure` to take out mentions of the `component` prop.

mnajdova commented:
> @mnajdova I renamed the ### Slot props section to ### Custom structure and updated the text. I'm not married to the header title, so I'm open to suggestions.

Perfect, thanks a lot Sam!

> This would need to be implemented across the rest of the Base docs as well, some of which also discuss slotProps in this section. But it feels confusing to talk about "the slot prop" and "the slotProps prop" in a section called "Slot props." 😅

Agree, we we are removing the `component` prop, we can do the same in the PRs.

mnajdova commented:
I will hold-off this PR till we have the Unstyled suffix removed, to avoid creating merge conflicts there.

mnajdova commented:
@samuelsycamore do we want to still keep the "Unstyled" in the title for the base pages, for e.g. https://deploy-preview-36677--material-ui.netlify.app/base/react-button/ vs https://mui.com/base/react-button/. I would vote for keeping only Button, as we are talking about the concept of button that is implemented trough component and hook.

hbjORbj commented on packages/api-docs-builder/ApiBuilders/ComponentApiBuilder.ts:
```diff
@@ -682,22 +685,25 @@ export default async function generateComponentApi(
   console.log('Built API docs for', reactApi.name);
 
+  const normalizedApiPathname = reactApi.apiPathname.replace(/\\/g, '/');
+  const noramlizedFilename = reactApi.filename.replace(/\\/g, '/');
```
addressed

hbjORbj commented on docs/data/base/components/button/button.md:
````diff
@@ -54,30 +54,44 @@ The Button component is composed of a root `<button>` slot with no interior slot
+The same applies for props specific to custom primitive elements:
+
+```tsx
+<Button<'svg'> viewBox="0 0 24 24" />
````
addressed

hbjORbj commented:
I think this should no longer appear.

`mui-base/src/Button/Button.spec.tsx`
![Screenshot 2023-04-25 at 12 15 24](https://user-images.githubusercontent.com/32841130/234260385-3ded4b50-c275-4bf5-8852-4565a0d7bdda.png)

michaldudak commented:
@hbjORbj, you may have your TS definitions cached from another branch. Restart the TS server, and the error should be gone.

hbjORbj commented:
@michaldudak Yes it's gone ;) thanks

hbjORbj reviewed (approved):
looks good!

samuelsycamore commented:
> @samuelsycamore do we want to still keep the "Unstyled" in the title for the base pages, for e.g. https://deploy-preview-36677--material-ui.netlify.app/base/react-button/ vs https://mui.com/base/react-button/. I would vote for keeping only Button, as we are talking about the concept of button that is implemented trough component and hook.

Agreed, I think it's better to drop "Unstyled" from the name wherever it comes up. I'd much rather teach our users to call it a "Base UI Button" than an "Unstyled Button."

hbjORbj commented on packages/mui-base/src/Button/Button.spec.tsx:
```diff
@@ -24,24 +24,31 @@ const polymorphicComponentTest = () => {
-      <Button component="a" href="#" />
+      <Button slots={{ root: 'a' }} href="#" />
 
-      <Button component={CustomComponent} stringProp="test" numberProp={0} />
```
I think we should keep this?

michaldudak replied:
Yup, it should stay (with `slots` instead of `component`, that is).

hbjORbj commented on docs/data/base/guides/overriding-component-structure/overriding-component-structure.md:
```diff
@@ -13,73 +13,38 @@ Slots are most commonly filled by HTML tags, but may also be filled by React com
 These slots are often (but not necessarily) nested within the root.
 
-For example, the [Slider](/base/react-slider/) is composed of a root `<span>` that houses several interior slots named for the elements they represent: track, thumb, rail, and so on.
+For example, the [Unstyled Slider](/base/react-slider/) is composed of a root `<span>` that houses several interior slots named for the elements they represent: track, thumb, rail, and so on.
```
```suggestion
For example, the [Base UI Slider](/base/react-slider/) is composed of a root `<span>` that houses several interior slots named for the elements they represent: track, thumb, rail, and so on.
```

hbjORbj commented on docs/data/base/guides/overriding-component-structure/overriding-component-structure.md:
```diff
@@ -13,73 +13,38 @@ Slots are most commonly filled by HTML tags, but may also be filled by React com
 You can use it to define additional custom props to pass to a component's interior slots.
 
-For example, the code snippet below shows how to add a custom CSS class to the badge slot of the [Base UI Badge](/base/react-badge/) component:
+For example, the code snippet below shows how to add a custom CSS class to the badge slot of the [Unstyled Badge](/base/react-badge/) component:
```
```suggestion
For example, the code snippet below shows how to add a custom CSS class to the badge slot of the [Base UI Badge](/base/react-badge/) component:
```

hbjORbj commented:
I made the final touches. Will merge it once CIs all pass.
